packages/cocoapods-downloader/src/cocoapods-downloader/base.ts

```diff
@@ -0,0 +1,153 @@
+import { mkdirSync } from "node:fs";
+
+import { version as projectVersion } from "@repo/cocoapods/package.json" with { type: "json" };
+import { execa, execaSync } from "execa";
+
+import { VERSION } from "./gem_version.js";
+
+export type UICallbacks = typeof defaultCallbacks;
+
+export abstract class Base<Options extends Record<string, unknown>> {
+  protected constructor(
+    public target_path: string,
+    public url: string,
+    public options: Partial<Options>,
+    private callbacks: UICallbacks = defaultCallbacks,
+  ) {
+    // TODO: Validate no unrecognized options?
+  }
+
+  get downloader_options(): Array<string> {
+    return [];
+  }
```

Comment:
Began writing this on stream. WIP:

```suggestion
export abstract class Base<Options extends Record<string, string>> {
  #options: Partial<Options>;
  static options(): Array<string> {
    return [];
  }
  get options(): Array<keyof Options> {
    return Object.keys(this.#options);
  }

  protected constructor(
    public target_path: string,
    public url: string,
    options: Partial<Options>,
    private callbacks: UICallbacks = defaultCallbacks,
  ) {
    this.#options = options;

    const supportedOptions = new Set(
      (this.constructor as typeof Base).options(),
    );
    const unrecognized_options = Object.keys(options).filter(
      (option) => !supportedOptions.has(option),
    );
    if (unrecognized_options.length > 0) {
      throw new Error(
        `Unrecognized options ${JSON.stringify(unrecognized_options)}`,
      );
    }
  }
```